src/security/CertificateProperties.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_SECURITY_CERTIFICATEPROPERTIES_H
+#define SQUID_SRC_SECURITY_CERTIFICATEPROPERTIES_H
+
+#include "security/forward.h"
+
+namespace Security
+{
+
+/**
+ * Certificate generation parameters to mimic
+ */
+class CertificateProperties
+{
+public:
+    CertificateProperties() = default;
+
+    Security::CertPointer mimicCert;          ///< Certificate to mimic
+
+    Security::CertPointer signWithX509;       ///< Certificate to sign the generated request
+    Security::PrivateKeyPointer signWithPkey; ///< The key of the signing certificate
+
+    bool setValidAfter = false;  ///< Do not mimic "Not Valid After" field
+    bool setValidBefore = false; ///< Do not mimic "Not Valid Before" field
+
+    bool setCommonName = false; ///< Replace the CN field of the mimicing subject with the given
+    std::string commonName;     ///< A CN to use for the generated certificate
+
+    CertSignAlgorithm signAlgorithm = Security::algSignEnd; ///< The signing algorithm to use
+    DigestAlgorithm signHash = nullptr;                     ///< The signing hash to use
+
+private:
+    CertificateProperties(CertificateProperties &) = delete;
```

Comment:
fixed that pre-existing const bug.

The undefined nature of these constructor and operator is pre-existing design for this class. I dont yet know enough about the ssl-bump code using this object to know whether removal of move is safe or not. So leaving that part out of scope for this change for now.

Reply:
> I dont yet know enough about the ssl-bump code using this object to know whether removal of move is safe or not.

Deleting a move constructor disables copying/moving of any kind. That is exactly what the old code was trying to do. You do not need to know anything about the SslBump code to improve/fix this.

Reply:
Why would move need to be forbidden? I can understand copy prevention to resolve issues with duplicate information. But prohibiting move breaks all non-reference temporary instance passing (and reference to temporary is forbidden).

Reply:
> Why would move need to be forbidden?

First of all, please note that in the current PR code, CertificateProperties moving is already forbidden (implicitly). For example:

```C++
static_assert(std::is_move_constructible<Security::CertificateProperties>::value, "CertificateProperties is move-constructible");
```

```
error: static assertion failed: CertificateProperties is move-constructible
```

I suggest _explicitly_ deleting the move constructor for many reasons, including these:

* It solves (some of the) CertificateProperties problems that this PR should solve.
* It is trivial.
* It is customary in Squid code to express "no copying of any kind" intent by deleting the move constructor (and nothing else).
* Explicitly deleting the copy constructor but not the move constructor raises questions that the code does not answer. We would have to add a complicated comment to explain ourselves.
* Both the official code and this PR code explicitly prohibit CertificateProperties copying.
* The corresponding move optimization is implicitly prohibited in the official code.
* The corresponding move optimization is implicitly prohibited in the PR code.

Please note that I suggest _explicitly_ deleting the move constructor without explicitly saying anything about the copy constructor or the assignment operator because, as we have discussed in the past, a modern way of prohibiting copying of any kind (including moving) is deleting the move constructor. If you insist on explicitly deleting all three copy/move members, I can live with that too. There are similar inconsistencies in Squid code already.


> But prohibiting move breaks all non-reference temporary instance passing

Please do not misinterpret the following comment as an agreement with the above statement, but if you want to add support for CertificateProperties moving, you can add that support. The current PR does not attempt to do that. AFAICT, neither the official code nor this PR support CertificateProperties moving. I suggest to keep things simple and continue to prohibit copying and moving until they are actually needed.

Reply:
Done in e9bf27786